File: App/Private Messages/MessageComposeViewController.m
//  MessageComposeViewController.m
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "MessageComposeViewController.h"
#import "AwfulAppDelegate.h"
#import "AwfulForumsClient.h"
#import "AwfulFrameworkCategories.h"
#import "AwfulThreadTagLoader.h"
#import "AwfulThreadTagPickerController.h"
#import "Awful-Swift.h"

@interface MessageComposeViewController () <AwfulThreadTagPickerControllerDelegate, UIViewControllerRestoration>

@property (strong, nonatomic) NewPrivateMessageFieldView *fieldView;

@property (strong, nonatomic) ThreadTag *threadTag;
@property (strong, nonatomic) AwfulThreadTagPickerController *threadTagPicker;

@property (assign, nonatomic) BOOL updatingThreadTags;
@property (copy, nonatomic) NSArray *availableThreadTags;

@end

@implementation MessageComposeViewController

- (instancetype)initWithRecipient:(User *)recipient
{
    if ((self = [self initWithNibName:nil bundle:nil])) {
        _recipient = recipient;
    }
    return self;
}

- (instancetype)initWithRegardingMessage:(PrivateMessage *)regardingMessage initialContents:(NSString *)initialContents
{
    if ((self = [self initWithNibName:nil bundle:nil])) {
        _regardingMessage = regardingMessage;
        _initialContents = [initialContents copy];
    }
    return self;
}

- (instancetype)initWithForwardingMessage:(PrivateMessage *)forwardingMessage initialContents:(NSString *)initialContents
{
    if ((self = [self initWithNibName:nil bundle:nil])) {
        _forwardingMessage = forwardingMessage;
        _initialContents = [initialContents copy];
    }
    return self;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    if ((self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil])) {
        self.title = @"Private Message";
        self.submitButtonItem.title = @"Send";
        self.restorationClass = self.class;
    }
    return self;
}

- (void)setThreadTag:(ThreadTag *)threadTag
{
    _threadTag = threadTag;
    [self updateThreadTagButtonImage];
}

- (void)loadView
{
    [super loadView];
    self.customView = self.fieldView;
}

- (void)themeDidChange
{
    [super themeDidChange];
    self.fieldView.toField.textField.textColor = self.textView.textColor;
    self.fieldView.toField.textField.keyboardAppearance = self.textView.keyboardAppearance;
    self.fieldView.subjectField.textField.textColor = self.textView.textColor;
    self.fieldView.subjectField.textField.keyboardAppearance = self.textView.keyboardAppearance;
    
    NSDictionary *styleAttrs = @{NSForegroundColorAttributeName: self.theme[@"placeholderTextColor"]};
    NSAttributedString *toString = [[NSAttributedString alloc] initWithString:@"To" attributes:styleAttrs];
    self.fieldView.toField.textField.attributedPlaceholder = toString;
    NSAttributedString *subjectString = [[NSAttributedString alloc] initWithString:@"Subject" attributes:styleAttrs];
    self.fieldView.subjectField.textField.attributedPlaceholder = subjectString;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self updateThreadTagButtonImage];
    if (self.recipient) {
        if (self.fieldView.toField.textField.text.length == 0) {
            self.fieldView.toField.textField.text = self.recipient.username;
        }
    } else if (self.regardingMessage) {
        if (self.fieldView.toField.textField.text.length == 0) {
            self.fieldView.toField.textField.text = self.regardingMessage.from.username;
        }
        if (self.fieldView.subjectField.textField.text.length == 0) {
            NSString *subject = self.regardingMessage.subject;
            if (![subject hasPrefix:@"Re: "]) {
                subject = [NSString stringWithFormat:@"Re: %@", subject];
            }
            self.fieldView.subjectField.textField.text = subject;
        }
        if (self.textView.text.length == 0) {
            self.textView.text = self.initialContents;
        }
    } else if (self.forwardingMessage) {
        if (self.fieldView.subjectField.textField.text.length == 0) {
            self.fieldView.subjectField.textField.text = [NSString stringWithFormat:@"Fw: %@", self.forwardingMessage.subject];
        }
        if (self.textView.text.length == 0) {
            self.textView.text = self.initialContents;
        }
    }
    
    [self updateAvailableThreadTagsIfNecessary];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self updateAvailableThreadTagsIfNecessary];
}

- (void)updateAvailableThreadTagsIfNecessary
{
    if (self.availableThreadTags.count > 0 || self.updatingThreadTags) return;
    self.updatingThreadTags = YES;
    __weak __typeof__(self) weakSelf = self;
    [[AwfulForumsClient client] listAvailablePrivateMessageThreadTagsAndThen:^(NSError *error, NSArray *threadTags) {
        __typeof__(self) self = weakSelf;
        self.availableThreadTags = threadTags;
        self.updatingThreadTags = NO;
    }];
}

- (NewPrivateMessageFieldView *)fieldView
{
    if (_fieldView) return _fieldView;
    _fieldView = [[NewPrivateMessageFieldView alloc] initWithFrame:CGRectMake(0, 0, 0, 88)];
    _fieldView.toField.label.textColor = [UIColor grayColor];
    _fieldView.subjectField.label.textColor = [UIColor grayColor];
    [_fieldView.threadTagButton addTarget:self
                                   action:@selector(didTapThreadTagButton:)
                         forControlEvents:UIControlEventTouchUpInside];
    [_fieldView.toField.textField addTarget:self
                                     action:@selector(toFieldDidChange)
                           forControlEvents:UIControlEventEditingChanged];
    [_fieldView.subjectField.textField addTarget:self
                                          action:@selector(subjectFieldDidChange)
                                forControlEvents:UIControlEventEditingChanged];
    return _fieldView;
}

- (void)updateThreadTagButtonImage
{
    UIImage *image;
    if (self.threadTag.imageName) {
        image = [AwfulThreadTagLoader imageNamed:self.threadTag.imageName];
    } else {
        image = [AwfulThreadTagLoader unsetThreadTagImage];
    }
    [self.fieldView.threadTagButton setImage:image forState:UIControlStateNormal];
}

- (void)didTapThreadTagButton:(ThreadTagButton *)button
{
    // TODO better handle waiting for available thread tags to download.
    if (self.availableThreadTags.count == 0) return;
    
    NSString *selectedImageName = self.threadTag.imageName ?: AwfulThreadTagLoaderEmptyPrivateMessageImageName;
    [self.threadTagPicker selectImageName:selectedImageName];
    [self.threadTagPicker presentFromView:button];
    
    // HACK: Calling -endEditing: once doesn't work if the To or Subject fields are selected. But twice works. I assume this is some weirdness from adding text fields as subviews to a text view.
    [self.view endEditing:YES];
    [self.view endEditing:YES];
}

- (AwfulThreadTagPickerController *)threadTagPicker
{
    if (_threadTagPicker) return _threadTagPicker;
    if (self.availableThreadTags.count == 0) return nil;
    
    NSMutableArray *imageNames = [NSMutableArray arrayWithObject:AwfulThreadTagLoaderEmptyPrivateMessageImageName];
    [imageNames addObjectsFromArray:[self.availableThreadTags valueForKey:@"imageName"]];
    _threadTagPicker = [[AwfulThreadTagPickerController alloc] initWithImageNames:imageNames secondaryImageNames:nil];
    _threadTagPicker.delegate = self;
    _threadTagPicker.navigationItem.leftBarButtonItem = _threadTagPicker.cancelButtonItem;
    return _threadTagPicker;
}

- (void)toFieldDidChange
{
    [self updateSubmitButtonItem];
}

- (void)subjectFieldDidChange
{
    [self updateSubmitButtonItem];
}

- (BOOL)canSubmitComposition
{
    return ([super canSubmitComposition] &&
            self.fieldView.toField.textField.text.length > 0 &&
            self.fieldView.subjectField.textField.text.length > 0);
}

- (NSString *)submissionInProgressTitle
{
    return @"Sending…";
}

- (void)submitComposition:(NSString *)composition completionHandler:(void (^)(BOOL))completionHandler
{
    __weak __typeof__(self) weakSelf = self;
    [[AwfulForumsClient client] sendPrivateMessageTo:self.fieldView.toField.textField.text
                                         withSubject:self.fieldView.subjectField.textField.text
                                           threadTag:self.threadTag
                                              BBcode:composition
                                    asReplyToMessage:self.regardingMessage
                                forwardedFromMessage:self.forwardingMessage
                                             andThen:^(NSError *error)
    {
        __typeof__(self) self = weakSelf;
        if (error) {
            completionHandler(NO);
            [self presentViewController:[UIAlertController alertWithNetworkError:error] animated:YES completion:nil];
        } else {
            completionHandler(YES);
        }
    }];
}

#pragma mark - AwfulThreadTagPickerControllerDelegate

- (void)threadTagPicker:(AwfulThreadTagPickerController *)picker didSelectImageName:(NSString *)imageName
{
    if ([imageName isEqualToString:AwfulThreadTagLoaderEmptyPrivateMessageImageName]) {
        self.threadTag = nil;
    } else {
        [self.availableThreadTags enumerateObjectsUsingBlock:^(ThreadTag *threadTag, NSUInteger i, BOOL *stop) {
            if ([threadTag.imageName isEqualToString:imageName]) {
                self.threadTag = threadTag;
                *stop = YES;
            }
        }];
    }
    [picker dismiss];
    [self focusInitialFirstResponder];
}

#pragma mark - UIViewControllerRestoration

+ (UIViewController *)viewControllerWithRestorationIdentifierPath:(NSArray *)identifierComponents coder:(NSCoder *)coder
{
    // AwfulObjectKey was introduced in Awful 3.2.
    UserKey *recipientKey = [coder decodeObjectForKey:RecipientUserKeyKey];
    if (!recipientKey) {
        NSString *recipientUserID = [coder decodeObjectForKey:obsolete_RecipientUserIDKey];
        if (recipientUserID) {
            recipientKey = [[UserKey alloc] initWithUserID:recipientUserID username:nil];
        }
    }
    PrivateMessageKey *regardingKey = [coder decodeObjectForKey:RegardingMessageKeyKey];
    if (!regardingKey) {
        NSString *regardingMessageID = [coder decodeObjectForKey:obsolete_RegardingMessageIDKey];
        if (regardingMessageID) {
            regardingKey = [[PrivateMessageKey alloc] initWithMessageID:regardingMessageID];
        }
    }
    PrivateMessageKey *forwardingKey = [coder decodeObjectForKey:ForwardingMessageKeyKey];
    if (!forwardingKey) {
        NSString *forwardingMessageID = [coder decodeObjectForKey:obsolete_ForwardingMessageIDKey];
        if (forwardingMessageID) {
            forwardingKey = [[PrivateMessageKey alloc] initWithMessageID:forwardingMessageID];
        }
    }
    NSString *initialContents = [coder decodeObjectForKey:InitialContentsKey];
    NSManagedObjectContext *managedObjectContext = [AwfulAppDelegate instance].managedObjectContext;
    MessageComposeViewController *newPrivateMessageViewController;
    if (recipientKey) {
        User *recipient = [User objectForKey:recipientKey inManagedObjectContext:managedObjectContext];
        newPrivateMessageViewController = [[MessageComposeViewController alloc] initWithRecipient:recipient];
    } else if (regardingKey) {
        PrivateMessage *regardingMessage = [PrivateMessage objectForKey:regardingKey inManagedObjectContext:managedObjectContext];
        newPrivateMessageViewController = [[MessageComposeViewController alloc] initWithRegardingMessage:regardingMessage
                                                                                         initialContents:initialContents];
    } else if (forwardingKey) {
        PrivateMessage *forwardingMessage = [PrivateMessage objectForKey:forwardingKey inManagedObjectContext:managedObjectContext];
        newPrivateMessageViewController = [[MessageComposeViewController alloc] initWithForwardingMessage:forwardingMessage
                                                                                          initialContents:initialContents];
    } else {
        newPrivateMessageViewController = [[MessageComposeViewController alloc] initWithRecipient:nil];
    }
    newPrivateMessageViewController.restorationIdentifier = identifierComponents.lastObject;
    return newPrivateMessageViewController;
}

- (void)encodeRestorableStateWithCoder:(NSCoder *)coder
{
    [super encodeRestorableStateWithCoder:coder];
    [coder encodeObject:self.recipient.objectKey forKey:RecipientUserKeyKey];
    [coder encodeObject:self.regardingMessage.objectKey forKey:RegardingMessageKeyKey];
    [coder encodeObject:self.forwardingMessage.objectKey forKey:ForwardingMessageKeyKey];
    [coder encodeObject:self.initialContents forKey:InitialContentsKey];
    [coder encodeObject:self.threadTag.objectKey forKey:ThreadTagKeyKey];
}

- (void)decodeRestorableStateWithCoder:(NSCoder *)coder
{
    self.fieldView.toField.textField.text = [coder decodeObjectForKey:ToKey];
    self.fieldView.subjectField.textField.text = [coder decodeObjectForKey:SubjectKey];
    // AwfulObjectKey was introduced in Awful 3.2.
    ThreadTagKey *tagKey = [coder decodeObjectForKey:ThreadTagKeyKey];
    if (!tagKey) {
        NSString *threadTagImageName = [coder decodeObjectForKey:obsolete_ThreadTagImageNameKey];
        if (threadTagImageName) {
            tagKey = [[ThreadTagKey alloc] initWithImageName:threadTagImageName threadTagID:nil];
        }
    }
    if (tagKey) {
        NSManagedObjectContext *managedObjectContext = (self.recipient.managedObjectContext ?:
                                                        self.regardingMessage.managedObjectContext ?:
                                                        self.forwardingMessage.managedObjectContext);
        self.threadTag = [ThreadTag objectForKey:tagKey inManagedObjectContext:managedObjectContext];
    }
    
    [super decodeRestorableStateWithCoder:coder];
}

static NSString * const RecipientUserKeyKey = @"RecipientUserKey";
static NSString * const obsolete_RecipientUserIDKey = @"AwfulRecipientUserID";
static NSString * const RegardingMessageKeyKey = @"RegardingMessageKey";
static NSString * const obsolete_RegardingMessageIDKey = @"AwfulRegardingMessageID";
static NSString * const ForwardingMessageKeyKey = @"ForwardingMessageKey";
static NSString * const obsolete_ForwardingMessageIDKey = @"AwfulForwardingMessageID";
static NSString * const InitialContentsKey = @"AwfulInitialContents";
static NSString * const ToKey = @"AwfulTo";
static NSString * const SubjectKey = @"AwfulSubject";
static NSString * const ThreadTagKeyKey = @"ThreadTagKey";
static NSString * const obsolete_ThreadTagImageNameKey = @"AwfulThreadTagImageName";

@end

File: App/Private Messages/MessageComposeViewController.swift
//  MessageComposeViewController.swift
//
//  Copyright 2016 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

import AwfulCore

/// For writing private messages.
final class MessageComposeViewController: ComposeTextViewController {
    fileprivate let recipient: User?
    fileprivate let regardingMessage: PrivateMessage?
    fileprivate let forwardingMessage: PrivateMessage?
    fileprivate let initialContents: String?
    fileprivate var threadTag: ThreadTag? {
        didSet { updateThreadTagButtonImage() }
    }
    fileprivate var availableThreadTags: [ThreadTag]?
    fileprivate var updatingThreadTags = false
    fileprivate var threadTagPicker: ThreadTagPickerViewController?
    
    fileprivate lazy var fieldView: NewPrivateMessageFieldView = {
        let fieldView = NewPrivateMessageFieldView(frame: CGRect(x: 0, y: 0, width: 0, height: 88))
        fieldView.toField.label.textColor = .gray
        fieldView.subjectField.label.textColor = .gray
        fieldView.threadTagButton.addTarget(self, action: #selector(didTapThreadTagButton), for: .touchUpInside)
        fieldView.toField.textField.addTarget(self, action: #selector(toFieldDidChange), for: .editingChanged)
        fieldView.subjectField.textField.addTarget(self, action: #selector(subjectFieldDidChange), for: .editingChanged)
        return fieldView
    }()
    
    init() {
        recipient = nil
        regardingMessage = nil
        forwardingMessage = nil
        initialContents = nil
        super.init(nibName: nil, bundle: nil)
        commonInit()
    }
    
    init(recipient: User) {
        self.recipient = recipient
        regardingMessage = nil
        forwardingMessage = nil
        initialContents = nil
        super.init(nibName: nil, bundle: nil)
        commonInit()
    }
    
    init(regardingMessage: PrivateMessage, initialContents: String?) {
        self.regardingMessage = regardingMessage
        recipient = nil
        forwardingMessage = nil
        self.initialContents = initialContents
        super.init(nibName: nil, bundle: nil)
        commonInit()
    }
    
    init(forwardingMessage: PrivateMessage, initialContents: String?) {
        self.forwardingMessage = forwardingMessage
        recipient = nil
        regardingMessage = nil
        self.initialContents = initialContents
        super.init(nibName: nil, bundle: nil)
        commonInit()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    fileprivate func commonInit() {
        title = "Private Message"
        submitButtonItem.title = "Send"
        restorationClass = type(of: self)
    }
    
    fileprivate func updateThreadTagButtonImage() {
        let image: UIImage
        if let
            imageName = threadTag?.imageName,
            let loadedImage = ThreadTagLoader.imageNamed(imageName)
        {
            image = loadedImage
        } else {
            image = ThreadTagLoader.unsetThreadTagImage
        }
        fieldView.threadTagButton.setImage(image, for: UIControlState())
    }
    
    fileprivate func updateAvailableThreadTagsIfNecessary() {
        guard availableThreadTags?.isEmpty ?? true else { return }
        guard !updatingThreadTags else { return }
        _ = ForumsClient.shared.listAvailablePrivateMessageThreadTags { [weak self] (error: Error?, threadTags: [ThreadTag]?) in
            self?.updatingThreadTags = false
            
            if let threadTags = threadTags {
                self?.availableThreadTags = threadTags
                
                let imageNames = [ThreadTagLoader.emptyPrivateMessageImageName] + threadTags.flatMap { $0.imageName }
                let picker = ThreadTagPickerViewController(imageNames: imageNames, secondaryImageNames: nil)
                picker.delegate = self
                picker.navigationItem.leftBarButtonItem = picker.cancelButtonItem
                self?.threadTagPicker = picker
            }
        }
    }
    
    override var canSubmitComposition: Bool {
        return super.canSubmitComposition &&
        fieldView.toField.textField.text?.isEmpty == false &&
        fieldView.subjectField.textField.text?.isEmpty == false
    }
    
    override var submissionInProgressTitle: String {
        return "Sending…"
    }
    
    override func submit(_ composition: String, completion: @escaping (Bool) -> Void) {
        guard let
            to = fieldView.toField.textField.text,
            let subject = fieldView.subjectField.textField.text
            else { return }
        let _ = ForumsClient.shared.sendPrivateMessage(to: to, subject: subject, threadTag: threadTag, bbcode: composition, regarding: regardingMessage, forwarding: forwardingMessage) { [weak self] (error: Error?) in
            if let error = error {
                completion(false)
                self?.present(UIAlertController.alertWithNetworkError(error), animated: true, completion: nil)
            } else {
                completion(true)
            }
        }
    }
    
    @objc fileprivate func didTapThreadTagButton(_ sender: ThreadTagButton) {
        guard let picker = threadTagPicker else { return }
        
        let selectedImageName = threadTag?.imageName ?? ThreadTagLoader.emptyPrivateMessageImageName
        picker.selectImageName(selectedImageName)
        picker.present(fromView: sender)
        
        // HACK: Calling -endEditing: once doesn't work if the To or Subject fields are selected. But twice works. I assume this is some weirdness from adding text fields as subviews to a text view.
        view.endEditing(true)
        view.endEditing(true)
    }
    
    @objc fileprivate func toFieldDidChange() {
        updateSubmitButtonItem()
    }
    
    @objc fileprivate func subjectFieldDidChange() {
        updateSubmitButtonItem()
    }
    
    // MARK: View lifecycle
    
    override func loadView() {
        super.loadView()
        customView = fieldView
    }
    
    override func themeDidChange() {
        super.themeDidChange()
        
        fieldView.toField.textField.textColor = textView.textColor
        fieldView.toField.textField.keyboardAppearance = textView.keyboardAppearance
        fieldView.subjectField.textField.textColor = textView.textColor
        fieldView.subjectField.textField.keyboardAppearance = textView.keyboardAppearance
        
        let attributes = [NSForegroundColorAttributeName: (theme["placeholderTextColor"] as UIColor?) ?? .gray]
        fieldView.toField.textField.attributedPlaceholder = NSAttributedString(string: "To", attributes: attributes)
        fieldView.subjectField.textField.attributedPlaceholder = NSAttributedString(string: "Subject", attributes: attributes)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        updateThreadTagButtonImage()
        
        if let recipient = recipient {
            if fieldView.toField.textField.text?.isEmpty ?? true {
                fieldView.toField.textField.text = recipient.username
            }
        } else if let regardingMessage = regardingMessage {
            if fieldView.toField.textField.text?.isEmpty ?? true {
                fieldView.toField.textField.text = regardingMessage.from?.username
            }
            
            if fieldView.subjectField.textField.text?.isEmpty ?? true {
                var subject = regardingMessage.subject ?? ""
                if !subject.hasPrefix("Re: ") {
                    subject = "Re: \(subject)"
                }
                fieldView.subjectField.textField.text = subject
            }
            
            if let initialContents = initialContents , textView.text.isEmpty {
                textView.text = initialContents
            }
        } else if let forwardingMessage = forwardingMessage {
            if fieldView.subjectField.textField.text?.isEmpty ?? true {
                fieldView.subjectField.textField.text = "Fw: \(forwardingMessage.subject ?? "")"
            }
            
            if let initialContents = initialContents , textView.text.isEmpty {
                textView.text = initialContents
            }
        }
        
        updateAvailableThreadTagsIfNecessary()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        updateAvailableThreadTagsIfNecessary()
    }
    
    // MARK: State restoration
    
    override func encodeRestorableState(with coder: NSCoder) {
        super.encodeRestorableState(with: coder)
        
        coder.encode(recipient?.objectKey, forKey: Keys.RecipientUserKey.rawValue)
        coder.encode(regardingMessage?.objectKey, forKey: Keys.RegardingMessageKey.rawValue)
        coder.encode(forwardingMessage?.objectKey, forKey: Keys.ForwardingMessageKey.rawValue)
        coder.encode(initialContents, forKey: Keys.InitialContents.rawValue)
        coder.encode(threadTag?.objectKey, forKey: Keys.ThreadTagKey.rawValue)
    }
    
    override func decodeRestorableState(with coder: NSCoder) {
        let context = AppDelegate.instance.managedObjectContext
        if let threadTagKey = coder.decodeObject(forKey: Keys.ThreadTagKey.rawValue) as? ThreadTagKey {
            threadTag = ThreadTag.objectForKey(objectKey: threadTagKey, inManagedObjectContext: context) as? ThreadTag
        }
        
        super.decodeRestorableState(with: coder)
    }
}

extension MessageComposeViewController: ThreadTagPickerViewControllerDelegate {
    func threadTagPicker(_ picker: ThreadTagPickerViewController, didSelectImageName imageName: String) {
        if imageName == ThreadTagLoader.emptyPrivateMessageImageName {
            threadTag = nil
        } else if let
            availableThreadTags = availableThreadTags,
            let i = availableThreadTags.index(where: { $0.imageName == imageName })
        {
            threadTag = availableThreadTags[i]
        }
        
        picker.dismiss()
        
        focusInitialFirstResponder()
    }
}

extension MessageComposeViewController: UIViewControllerRestoration {
    static func viewController(withRestorationIdentifierPath identifierComponents: [Any], coder: NSCoder) -> UIViewController? {
        let recipientKey = coder.decodeObject(forKey: Keys.RecipientUserKey.rawValue) as? UserKey
        let regardingKey = coder.decodeObject(forKey: Keys.RegardingMessageKey.rawValue) as? PrivateMessageKey
        let forwardingKey = coder.decodeObject(forKey: Keys.ForwardingMessageKey.rawValue) as? PrivateMessageKey
        let initialContents = coder.decodeObject(forKey: Keys.InitialContents.rawValue) as? String
        let context = AppDelegate.instance.managedObjectContext
        
        let composeViewController: MessageComposeViewController
        if let
            recipientKey = recipientKey,
            let recipient = User.objectForKey(objectKey: recipientKey, inManagedObjectContext: context) as? User
        {
            composeViewController = MessageComposeViewController(recipient: recipient)
        } else if let
            regardingKey = regardingKey,
            let regardingMessage = PrivateMessage.objectForKey(objectKey: regardingKey, inManagedObjectContext: context) as? PrivateMessage
        {
            composeViewController = MessageComposeViewController(regardingMessage: regardingMessage, initialContents: initialContents)
        } else if let
            forwardingKey = forwardingKey,
            let forwardingMessage = PrivateMessage.objectForKey(objectKey: forwardingKey, inManagedObjectContext: context) as? PrivateMessage
        {
            composeViewController = MessageComposeViewController(forwardingMessage: forwardingMessage, initialContents: initialContents)
        } else {
            return nil
        }
        composeViewController.restorationIdentifier = identifierComponents.last as? String
        return composeViewController
    }
}

private enum Keys: String {
    case RecipientUserKey
    case RegardingMessageKey
    case ForwardingMessageKey
    case InitialContents
    case ThreadTagKey
}
